src/strands/event_loop/event_loop.py

```diff
@@ -112,50 +117,68 @@ async def event_loop_cycle(agent: "Agent", kwargs: dict[str, Any]) -> AsyncGener
             model_id=model_id,
         )
 
+        get_registry(agent).invoke_callbacks(
+            BeforeModelInvocationEvent(
+                agent=agent,
+            )
+        )
+
         try:
             # TODO: To maintain backwards compatibility, we need to combine the stream event with kwargs before yielding
             #       to the callback handler. This will be revisited when migrating to strongly typed events.
-            async for event in stream_messages(agent.model, agent.system_prompt, agent.messages, agent.tool_config):
+            async for event in stream_messages(
+                agent.model, agent.system_prompt, agent.messages, tool_config=agent.tool_config
+            ):
                 if "callback" in event:
                     yield {"callback": {**event["callback"], **(kwargs if "delta" in event["callback"] else {})}}
 
             stop_reason, message, usage, metrics = event["stop"]
             kwargs.setdefault("request_state", {})
 
+            get_registry(agent).invoke_callbacks(
+                AfterModelInvocationEvent(
+                    agent=agent,
+                    stop_data=AfterModelInvocationEvent.ModelStopResponse(
+                        stop_reason=stop_reason,
+                        message=message,
+                    ),
+                )
+            )
+
             if model_invoke_span:
                 tracer.end_model_invoke_span(model_invoke_span, message, usage, stop_reason)
             break  # Success! Break out of retry loop
 
-        except ContextWindowOverflowException as e:
-            if model_invoke_span:
-                tracer.end_span_with_error(model_invoke_span, str(e), e)
-            raise e
-
-        except ModelThrottledException as e:
+        except Exception as e:
             if model_invoke_span:
                 tracer.end_span_with_error(model_invoke_span, str(e), e)
 
-            if attempt + 1 == MAX_ATTEMPTS:
-                yield {"callback": {"force_stop": True, "force_stop_reason": str(e)}}
-                raise e
-
-            logger.debug(
-                "retry_delay_seconds=<%s>, max_attempts=<%s>, current_attempt=<%s> "
-                "| throttling exception encountered "
-                "| delaying before next retry",
-                current_delay,
-                MAX_ATTEMPTS,
-                attempt + 1,
+            get_registry(agent).invoke_callbacks(
+                AfterModelInvocationEvent(
+                    agent=agent,
+                    exception=e,
+                )
             )
-            time.sleep(current_delay)
-            current_delay = min(current_delay * 2, MAX_DELAY)
 
-            yield {"callback": {"event_loop_throttled_delay": current_delay, **kwargs}}
+            if isinstance(e, ModelThrottledException):
```

Comment:
I think it would make sense that this could be re-implemented as a hook, but it requires additional design and considerations - especially around hook ordering and how to allow customers to supress the behavior